File: packages/fastmd-shared/src/protocol.ts
/**
 * NDJSON RPC protocol definitions for FastMD sidecar communication
 */

export const PROTOCOL_VERSION = '1.0.0';

export type RpcId = string | number;

export interface RpcRequest {
  jsonrpc: '2.0';
  id: RpcId;
  method: string;
  params?: unknown;
}

export interface RpcResponse {
  jsonrpc: '2.0';
  id: RpcId;
  result?: unknown;
  error?: RpcError;
}

export interface RpcError {
  code: number;
  message: string;
  data?: unknown;
}

export interface RpcNotification {
  jsonrpc: '2.0';
  method: string;
  params?: unknown;
}

// Standard error codes
export const RPC_ERROR_CODES = {
  PARSE_ERROR: -32700,
  INVALID_REQUEST: -32600,
  METHOD_NOT_FOUND: -32601,
  INVALID_PARAMS: -32602,
  INTERNAL_ERROR: -32603,
  // Custom error codes
  TRANSFORM_ERROR: -32001,
  CACHE_ERROR: -32002,
  IO_ERROR: -32003
} as const;

// Method names for sidecar operations
export const RPC_METHODS = {
  // Core operations
  TRANSFORM: 'transform',
  NORMALIZE: 'normalize',
  COMPUTE_DIGEST: 'computeDigest',

  // Cache operations
  CACHE_GET: 'cache.get',
  CACHE_SET: 'cache.set',
  CACHE_DELETE: 'cache.delete',
  CACHE_CLEAR: 'cache.clear',
  CACHE_STATS: 'cache.stats',

  // Lifecycle
  PING: 'ping',
  SHUTDOWN: 'shutdown',
  CONFIGURE: 'configure'
} as const;

// Request/Response types for each method
export interface TransformRequest {
  file: string;
  content: string;
  options?: {
    mode?: 'development' | 'production';
    sourcemap?: boolean;
    framework?: 'astro' | 'vite';
  };
}

export interface TransformResponse {
  code: string;
  map?: unknown;
  metadata?: Record<string, unknown>;
  dependencies?: string[];
}

export interface NormalizeRequest {
  content: string;
  removeBom?: boolean;
  normalizeLf?: boolean;
}

export interface NormalizeResponse {
  content: string;
  changed: boolean;
}

export interface ComputeDigestRequest {
  files: Array<{
    path: string;
    size: number;
    mtime: number;
  }>;
}

export interface ComputeDigestResponse {
  digest: string;
}

export interface CacheGetRequest {
  key: string;
}

export interface CacheGetResponse {
  found: boolean;
  entry?: {
    code: string;
    map?: unknown;
    metadata?: Record<string, unknown>;
  };
}

export interface CacheSetRequest {
  key: string;
  code: string;
  map?: unknown;
  metadata?: Record<string, unknown>;
}

export interface CacheSetResponse {
  success: boolean;
}

export interface CacheStatsResponse {
  entries: number;
  size: number;
  hits: number;
  misses: number;
  hitRate: number;
}

export interface ConfigureRequest {
  cacheDir?: string;
  logLevel?: 'silent' | 'info' | 'debug' | 'trace';
  features?: Record<string, boolean>;
}

export interface ConfigureResponse {
  success: boolean;
  applied: Record<string, unknown>;
}

File: engines/sidecar/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RpcMessage {
    Request(RpcRequest),
    Notification(RpcNotification),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RpcRequest {
    pub jsonrpc: String,
    pub id: RpcId,
    pub method: String,
    pub params: Option<Value>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RpcNotification {
    pub jsonrpc: String,
    pub method: String,
    pub params: Option<Value>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RpcResponse {
    pub jsonrpc: String,
    pub id: RpcId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<RpcError>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RpcId {
    Number(i64),
    String(String),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RpcError {
    pub code: i32,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<Value>,
}

// Error codes
pub const PARSE_ERROR: i32 = -32700;
#[allow(dead_code)]
pub const INVALID_REQUEST: i32 = -32600;
pub const METHOD_NOT_FOUND: i32 = -32601;
pub const INVALID_PARAMS: i32 = -32602;
#[allow(dead_code)]
pub const INTERNAL_ERROR: i32 = -32603;

// Custom error codes
pub const TRANSFORM_ERROR: i32 = -32001;
#[allow(dead_code)]
pub const CACHE_ERROR: i32 = -32002;
#[allow(dead_code)]
pub const IO_ERROR: i32 = -32003;

pub fn create_response(id: RpcId, result: Value) -> RpcResponse {
    RpcResponse {
        jsonrpc: "2.0".to_string(),
        id,
        result: Some(result),
        error: None,
    }
}

pub fn create_error_response(id: RpcId, code: i32, message: String, data: Option<Value>) -> RpcResponse {
    RpcResponse {
        jsonrpc: "2.0".to_string(),
        id,
        result: None,
        error: Some(RpcError {
            code,
            message,
            data,
        }),
    }
}

pub fn create_parse_error() -> RpcResponse {
    RpcResponse {
        jsonrpc: "2.0".to_string(),
        id: RpcId::String("null".to_string()),
        result: None,
        error: Some(RpcError {
            code: PARSE_ERROR,
            message: "Parse error".to_string(),
            data: None,
        }),
    }
}

pub fn create_method_not_found(id: RpcId) -> RpcResponse {
    create_error_response(id, METHOD_NOT_FOUND, "Method not found".to_string(), None)
}